app/controllers/api_controller/roles.rb

```diff
@@ -0,0 +1,81 @@
+class ApiController
+  module Roles
+    include FeatureHelper
+    def create_resource_roles(type, _id = nil, data = {})
+      if data.key?("id") || data.key?("href")
+        msg = "Resource id or href should not be specified for creating a new #{type}"
+        api_log_info("BadRequestError, #{msg}")
+        raise BadRequestError, msg
+      end
+
+      # Can't create a read-only role (reserved for out-of-box roles)
+      if data ['read_only']
+        msg = "Cannot create a read-only role."
+        api_log_info("BadRequestError, #{msg}")
+        raise BadRequestError, msg
+      end
+
+      role_klass = collection_class(type)
+
+      get_settings_and_features(data)
+
+      role = role_klass.create!(data.except(*ID_ATTRS))
+      api_log_info("Created new role #{role.name}")
+      role
+    rescue => err
+      role.destroy if role
+      raise err
+    end
+
+    def edit_resource_roles(type, id = nil, data = {})
+      unless id
+        msg = "Must specify an id for editing a #{type} resource"
+        api_log_info("BadRequestError, #{msg}")
```

Comment:
same as other comment, once error_handler is updated, you can remove all those api_log_info lines.